tests/tools/utils.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+import json
+import os
+import re
+import sys
+
+import boto3
+import botocore.client
+
+
+def resource_has_tags(resource, tags={}):
+    """Checks if a resource defines all `tags`."""
+    return all(
+        tag_key in resource.keys() and resource[tag_key] == tag_value
+        for (tag_key, tag_value) in tags.items()
+    )
+
+
+class S3ResourceFetcher:
+    """A class for fetching vmm-reference test resources from S3."""
+
+    def __init__(
+            self,
+            resource_bucket,
+            resource_manifest_path,
+            download_location=None,
+    ):
+        """Initialize S3 client, manifest of test resources and S3 bucket name."""
+        with open(resource_manifest_path) as json_file:
+            self._resource_manifest = json.load(json_file)
+        self._latest_version = self.get_latest_version()
+        self._resource_bucket = resource_bucket
+        self._s3 = boto3.client(
+            's3',
+            config=botocore.client.Config(signature_version=botocore.UNSIGNED)
+        )
+        if download_location is None:
+            path = os.path.join(
+                os.path.dirname(__file__),
+                "../../resources")
+            self._default_download_location = os.path.abspath(path)
+        else:
+            self._default_download_location = download_location
+
+        print(
+            "Setting default download location: ", self._default_download_location,
+            file=sys.stderr)
+
+    def download(self, resource_type, resource_name, tags={}, version=None, download_location=None):
+        """Downloads the resource specified by name and type. Optionally, the user can
+        specify what tags the resource needs to have, and its version.
+
+        The default version used when downloading is the latest version as defined in the
+        resource manifest."""
+        # TODO: The resource_name does not need to be required, because we can download
+        # resources by type & tags.
+        if version is None:
+            version = self._latest_version
+        if download_location is None:
+            download_location = self._default_download_location
```

Comment:
nit: We can simply use `_latest_version` and `_default_download_version` as defaults.

Reply:
I couldn't find a way to do that because they're defined in self. That's why I implemented this workaround.

Reply:
Yes, makes sense. Sorry I missed the `self` part.

Reply:
A slightly efficient (not that it matters) and somewhat readable way might be - 
`version = version or self.get_latest_version()` and similarly `download_location = download_location or self.get_download_location()`  (and then we do not have to initialize them in the constructor). This will also deal with someone passing "empty" strings accidentally.